File: Valid Sudoku.cpp
#include <unordered_set>
#define F(i,a,b) for( int i = a; i <= b; i++ )

class Solution {
public:
    unordered_set<char> c1;

    bool isValidSudoku(vector<vector<char> > &board) {
        //row
        F(i, 0, 8){
            F(j, 0, 8){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
            c1.clear();
        }

        //col
        F(j, 0, 8){
            F(i, 0, 8){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
            c1.clear();
        }

        //block
        F(j, 0, 2){
            F(i, 0, 2){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 0, 2){
            F(i, 3, 5){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 0, 2){
            F(i, 6, 8){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();

        F(j, 3, 5){
            F(i, 0, 2){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 3, 5){
            F(i, 3, 5){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 3, 5){
            F(i, 6, 8){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();

        F(j, 6, 8){
            F(i, 0, 2){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 6, 8){
            F(i, 3, 5){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();
        F(j, 6, 8){
            F(i, 6, 8){
                if( c1.find(board[i][j]) != c1.end() ){
                    return false;
                }
                if( board[i][j] != '.' ){ 
                    c1.insert(board[i][j]); 
                }
            }
        }
        c1.clear();

        return true;
    }
};

File: LeetCode-Java/Valid Sudoku.java




public class Solution {
	public boolean isValidSudoku(char[][] board) {

		for( int i = 0; i < 9; i++ ){
			boolean [] ht = new boolean [10];
			for( int j = 0; j < 9; j++ ){
				if( board[i][j] != '.' ){
					int tmp = board[i][j] - '0';
					if( ht[tmp] ){
						return false;
					}else{
						ht[tmp] = true;
					}
				}
			}
		}

		for( int j = 0; j < 9; j++ ){
			boolean [] ht = new boolean [10];
			for( int i = 0; i < 9; i++ ){
				if( board[i][j] != '.' ){
					int tmp = board[i][j] - '0';
					if( ht[tmp] ){
						return false;
					}else{
						ht[tmp] = true;
					}
				}
			}
		}

		for( int m = 0; m < 3; m++ ){
			for( int n = 0; n < 3; n++ ){
				boolean [] ht = new boolean [10];
				for( int i = 3*m; i < 3*m+3; i++ ){		
					for( int j = 3*n; j < 3*n+3; j++ ){
						if( board[i][j] != '.' ){
							int tmp = board[i][j] - '0';
							if( ht[tmp] ){
								return false;
							}else{
								ht[tmp] = true;
							}
						}
					}
				}
			}
		}

		return true;

	}
}



